tests/accessor/accessor_default_values.h

```diff
@@ -0,0 +1,174 @@
+/*******************************************************************************
+//
+//  SYCL 2020 Conformance Test Suite
+//
+//  This file provides functions for tests on accessor default values.
+//
+*******************************************************************************/
+#ifndef SYCL_CTS_ACCESSOR_DEFAULT_VALUES_H
+#define SYCL_CTS_ACCESSOR_DEFAULT_VALUES_H
+
+#include "accessor_common.h"
+
+#include <type_traits>
+
+namespace accessor_default_values_test {
+
+namespace detail {
+constexpr int expected_dims = 1;
+constexpr sycl::target expected_target = sycl::target::device;
+constexpr sycl::access::placeholder expected_placeholder =
+    sycl::access::placeholder::false_t;
+template <typename T>
+constexpr auto expected_mode =
+    std::is_const_v<T> ? sycl::access_mode::read : sycl::access_mode::read_write
+}  // namespace detail
+
+/**
+ * @brief Provides functor that lets verify that local_accessor's template
+ *        parameters have expected default values
+ * @tparam T Current data type
+ * @param type_name Current data type string representation
+ */
+template <typename T>
+class test_for_local_acc {
+ public:
+  void operator()(const std::string& type_name) {
+    auto section_name =
+        "Verify default value for dimensions template parameter for " +
+        type_name + " data type. [local_accessor]";
+    SECTION(section_name) {
+      REQUIRE(std::is_same_v<sycl::local_accessor<T>,
+                             sycl::local_accessor<T, detail::expected_dims>>);
+    }
+  }
+};
+
+/**
+ * @brief Provides functor that lets verify that host_accessor's template
+ *        parameters have expected default values
+ * @tparam T Current data type
+ * @param type_name Current data type string representation
+ */
+
+template <typename T, typename AccessModeT, typename DimensionT,
+          typename TargetT>
+class test_for_host_acc {
+  static constexpr sycl::access_mode AccessMode = AccessModeT::value;
+  static constexpr int Dimension = DimensionT::value;
+  static constexpr sycl::target Target = TargetT::value;
+
+ public:
+  void operator()(const std::string& type_name,
+                  const std::string& access_mode_name,
+                  const std::string& target_name) {
+    auto section_name =
+        get_section_name<Dimension>(type_name, access_mode_name, target_name,
+                                    "Verify default value for dimensions "
+                                    "template parameter. [host_accessor]");
+    SECTION(section_name) {
+      REQUIRE(std::is_same_v<sycl::host_accessor<T>,
+                             sycl::host_accessor<T, detail::expected_dims>>);
```

Comment:
We are not using `Dimension` here, but we would run this check for every `Dimension` value. Is it intended?
Maybe we might have something like the folllowing here?
```
constexpr bool should_check_dimension =
    (Dimension == detail::expected_dims);

if constexpr (should_check_dimension ) {
```

Reply:
this is by design because we may not iterate over default values of `Dimension`, `target`, e.t.c. and in this case this verifications will be skipped

Reply:
You are right, it definitely makes things much harder.

Still I'm bit worried that we don't have any simple logic to switch off repeating checks in such case.
For example, if `sycl::accessor<sycl::half>` doesn't have proper default dimension value for some reason - we don't need to print out an assertion failure for each possible combination of the rest template parameters.
How do you think, might something like this be handy?

```
namespace detail {
template <template <typename...> class, typename ...>
struct run_once_per {
  /**
   * @brief Returns true only once per application run for each instantiation of
   *        `run_once_per` class
   * @details Uses Meyer's singleton to have lazy initialization with no static
   *          initialization order fiasco possible; not thread-safe
   */
  static bool should_run() {
    static bool triggered = false;
    const bool result = triggered;
    return result ^ (triggered = true);
  }
};
}

template <typename T, typename AccessModeT, typename DimensionT,
          typename TargetT>
class test_for_host_acc {
  template <typename ... Types>
  using run_once_per = detail::run_once_per<test_for_host_acc, T, Types...>;
  using run_once = run_once_per<>;
  ...
  if (run_once::should_run()) {
    ...
    REQUIRE(std::is_same_v<sycl::host_accessor<T>,
                             sycl::host_accessor<T, detail::expected_dims>>);
  }
  ...
};

template <typename T, typename AccessModeT, typename DimensionT,
          typename TargetT>
class test_for_generic_acc {
  template <typename ... Types>
  using run_once_per = detail::run_once_per<test_for_generic_acc, T, Types...>;
  using run_once = run_once_per<>;
  ...
  if (run_once::should_run()) {
    ...
    REQUIRE(std::is_same_v<sycl::host_accessor<T>,
                             sycl::host_accessor<T, detail::expected_dims>>);
  }
  if (run_once_per<DimensionT>::should_run()) {
    ...
    REQUIRE(std::is_same_v<
              sycl::accessor<T, Dimension>,
              sycl::accessor<T, Dimension, detail::expected_mode<T>>>);
  }
  if (run_once_per<DimensionT, AccessModeT>::should_run()) {
    ...
    REQUIRE(std::is_same_v<sycl::accessor<T, Dimension, AccessMode>,
                             sycl::accessor<T, Dimension, AccessMode,
                                            detail::expected_target>>);
  }
  ...
};
```

Reply:
As a follow-up for the previous suggestion - looks like it would be better to state dependency on `T` explicitly:
```
template <typename T, typename AccessModeT, typename DimensionT,
          typename TargetT>
class test_for_generic_acc {
  template <typename ... Types>
  using run_once_per = detail::run_once_per<test_for_generic_acc, Types...>;
  ...
  if (run_once_per<T>::should_run()) {
    ...
    REQUIRE(std::is_same_v<sycl::host_accessor<T>,
                             sycl::host_accessor<T, detail::expected_dims>>);
  }
  if (run_once_per<T, DimensionT>::should_run()) {
    ...
    REQUIRE(std::is_same_v<
              sycl::accessor<T, Dimension>,
              sycl::accessor<T, Dimension, detail::expected_mode<T>>>);
  }
```

Reply:
> For example, if sycl::accessor<sycl::half> doesn't have proper default dimension value

This should be explicitly state into SYCL-2020 spec., until all specializations should have specified default values, so now we should werify it

> How do you think, might something like this be handy?

We may use such solution, but it looks some strange :)
We should add not a little code for skip some the same verifications, but this could be useful

Reply:
@andreyromanof @kopylovanat What do you think about this suggesion?

Reply:
I guess we can easily skip extra checks by using `if constexpr`.
Something like this:
```
if constexpr (Dimension == /*something*/) { //preform one-time check}
```
but I'm also okay with the current solution.